tools/rpkg/src/register.cpp

```diff
@@ -58,25 +63,149 @@ class RArrowTabularStreamFactory {
 	RArrowTabularStreamFactory(SEXP export_fun_p, SEXP arrow_scannable_p)
 	    : arrow_scannable(arrow_scannable_p), export_fun(export_fun_p) {};
 
-	static unique_ptr<ArrowArrayStreamWrapper> Produce(uintptr_t factory_p) {
-		RProtector r;
-		int err;
-		auto factory = (RArrowTabularStreamFactory *)factory_p;
+	static unique_ptr<ArrowArrayStreamWrapper>
+	Produce(uintptr_t factory_p, std::pair<std::unordered_map<idx_t, string>, std::vector<string>> &project_columns,
+	        TableFilterCollection *filters) {
 
+		RProtector r;
 		auto res = make_unique<ArrowArrayStreamWrapper>();
+		auto factory = (RArrowTabularStreamFactory *)factory_p;
 		auto stream_ptr_sexp =
 		    r.Protect(Rf_ScalarReal(static_cast<double>(reinterpret_cast<uintptr_t>(&res->arrow_array_stream))));
-		// export the arrow scannable to a stream pointer
-		auto export_call = r.Protect(Rf_lang3(factory->export_fun, factory->arrow_scannable, stream_ptr_sexp));
-		R_tryEval(export_call, R_GlobalEnv, &err);
-		if (err) {
-			Rf_error("Failed to produce arrow stream");
+		SEXP export_call;
+
+		auto export_fun = r.Protect(VECTOR_ELT(factory->export_fun, 0));
+		if (project_columns.second.empty()) {
+			export_call = r.Protect(Rf_lang3(export_fun, factory->arrow_scannable, stream_ptr_sexp));
+		} else {
+			auto projection_sexp = r.Protect(RApi::StringsToSexp(project_columns.second));
+			SEXP filters_sexp = r.Protect(Rf_ScalarLogical(true));
+			if (filters && filters->table_filters && !filters->table_filters->filters.empty()) {
+
+				filters_sexp = r.Protect(TransformFilter(*filters, project_columns.first, factory->export_fun));
+			}
+			export_call = r.Protect(
+			    Rf_lang5(export_fun, factory->arrow_scannable, stream_ptr_sexp, projection_sexp, filters_sexp));
 		}
+		RApi::REvalThrows(export_call);
 		return res;
 	}
 
 	SEXP arrow_scannable;
 	SEXP export_fun;
+
+private:
```

Comment:
If I read this correctly, it looks like you're reimplementing how to build Arrow `Expression` objects to pass as a filter. This (and many many more functions) is all handled in the `arrow` package already--can you call the same code that our `dplyr::filter()` method does instead?